### PR TITLE
[LPT] MoveFakeQuantize

### PR DESCRIPTION
**Pull request by NikDemoShow**
### Details:
 - MoveFakeQuantizeTransformation
 - Functional Tests
 - Plugin Tests

### Tickets:
 - *53934*
 - Validation: *62106*

### CODE REVIEW

**Comment by ilyachur** [comment]
Looks like you have a lot of redundant changes in this PR.

File: inference-engine/tests/ngraph_helpers/lpt_ngraph_functions/src/move_fake_quantize_function.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                const std::int64_t& axis) {
+
+                const auto input1 = std::make_shared<ngraph::opset1::Parameter>(inputPrecision, inputShape);
+                input1->set_friendly_name("input1");
+
+                const auto input2 = std::make_shared<ngraph::opset1::Parameter>(inputPrecision, inputShape);
+                input2->set_friendly_name("input2");
+
+
+                if (fqOnData3.empty()) {
```
Comment (eshoguli) [inline inference-engine/tests/ngraph_helpers/lpt_ngraph_functions/src/move_fake_quantize_function.cpp:46]: please refactor in this way:
```cpp
std::shared_ptr<Node> parent = ...
if (!fqOnData1.empty()) {
    parent = ...
}
if (!fqOnData2.empty()) {
    parent = ...
}
if (!fqOnData3.empty()) {
    parent = ...
}
```

File: inference-engine/tests/functional/inference_engine/lp_transformations/move_fake_quantize_for_concat_transformation.cpp [hunk 2]
```diff
@@ -0,0 +1,346 @@
+// Copyright (C) 2018-2021 Intel Corporation
```
Comment (eshoguli) [inline inference-engine/tests/functional/inference_engine/lp_transformations/move_fake_quantize_for_concat_transformation.cpp:1]: move_fake_quantize_for_concat_transformation.cpp => move_fake_quantize_transformation.cpp

File: inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                this->register_matcher(m, callback);
+            }
+
+            bool MoveFakeQuantize::transform(TransformationContext& context, ngraph::pattern::Matcher& m) {
+                auto fq = m.get_match_root();
+                auto result = *fq->output(0).get_target_inputs().begin();
+                auto operation = fq->get_input_node_shared_ptr(0);
+                auto type = operation->get_type_name();
+                std::shared_ptr<ngraph::Node> concat, fq1input, fq2input;
+                if (strcmp(type, "Concat") == 0) {
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp:42]: extend pattern matching & remove

File: inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/move_fake_quantize_transformation.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace {
+const std::vector<ngraph::element::Type> netPrecisions = {
+    ngraph::element::f32,
+    ngraph::element::f16
+};
+
+    const std::vector<ngraph::pass::low_precision::LayerTransformation::Params> trasformationParamValues = {
+       LayerTestsUtils::LayerTransformationParamsNGraphFactory::createParams(),
+       LayerTestsUtils::LayerTransformationParamsNGraphFactory::createParams().setUpdatePrecisions(false),
```
Comment (v-Golubev) [inline inference-engine/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/move_fake_quantize_transformation.cpp:20]: Could you please remove `LayerTestsUtils::LayerTransformationParamsNGraphFactory::createParams().setUpdatePrecisions(false)` ?

File: inference-engine/tests/functional/plugin/shared/src/low_precision_transformations/move_fake_quantize_transformation.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    std::ostringstream result;
+    result << getTestCaseNameByParams(netPrecision, inputShape, targetDevice, params) << "_" <<
+        param.operation;
+    return result.str();
+}
+
+void MoveFakeQuantizeTransformation::SetUp() {
+    threshold = 0.1f;
+
```
Comment (v-Golubev) [inline inference-engine/tests/functional/plugin/shared/src/low_precision_transformations/move_fake_quantize_transformation.cpp:37]: To my mind we can use default threshold in this tests. Please, remove

**Review by v-Golubev — COMMENTED** [review]
Please fix comments

**Review by v-Golubev — APPROVED** [review]
Look good to me, thanks!

File: inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            fqInput = std::make_shared<ngraph::opset1::Relu>(input);
+            fqInput->set_friendly_name(relu_original_name + std::to_string(i + 1));
+        }
+        auto newFq = std::make_shared<opset1::FakeQuantize>(fqInput,
+            fq->get_input_node_shared_ptr(1),
+            fq->get_input_node_shared_ptr(2),
+            fq->get_input_node_shared_ptr(3),
+            fq->get_input_node_shared_ptr(4),
+            as_type_ptr<opset1::FakeQuantize>(fq)->get_levels());
+        newFq->set_friendly_name(fq_original_name + std::to_string(i + 1));
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp:86]: add delimiter:
``` cpp
newFq->set_friendly_name(fq_original_name + "_" + std::to_string(i + 1));
```

File: inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        concat = operation;
+    } else {
+        relu_original_name = operation->get_friendly_name();
+        concat = operation->get_input_node_shared_ptr(0);
+        only_concat = false;
+    }
+    std::vector<std::shared_ptr<ngraph::Node>> fqs;
+    size_t input_size = concat->get_input_size();
+    for (size_t i{ 0 }; i < input_size; ++i) {
+        std::shared_ptr<ngraph::Node> fqInput;
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp:72]: one naming conversion: `fqInput` => `fq_input`

File: inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+            auto input = concat->get_input_node_shared_ptr(i);
+            fqInput = std::make_shared<ngraph::opset1::Relu>(input);
+            fqInput->set_friendly_name(relu_original_name + std::to_string(i + 1));
+        }
+        auto newFq = std::make_shared<opset1::FakeQuantize>(fqInput,
+            fq->get_input_node_shared_ptr(1),
+            fq->get_input_node_shared_ptr(2),
+            fq->get_input_node_shared_ptr(3),
+            fq->get_input_node_shared_ptr(4),
+            as_type_ptr<opset1::FakeQuantize>(fq)->get_levels());
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp:85]: use `clone_with_new_inputs` here

File: inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    }
+    ngraph::copy_runtime_info(fq, fqs);
+    auto get_outputs = [](const std::vector<std::shared_ptr<ngraph::Node>>& fqs) -> const ngraph::OutputVector {
+        ngraph::OutputVector result;
+        for (auto& fq : fqs) {
+            result.push_back(fq->output(0));
+        }
+        return result;
+    };
+    auto newConcat = concat->clone_with_new_inputs(get_outputs(fqs));
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp:97]: don't need `get_outputs`:
``` cpp
auto newConcat = concat->clone_with_new_inputs(ngraph::OutputVector(fqs.begin(), fqs.end()));
```

File: inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    auto fq2 = std::make_shared<opset1::FakeQuantize>(fq2input,
+        fq->get_input_node_shared_ptr(1),
+        fq->get_input_node_shared_ptr(2),
+        fq->get_input_node_shared_ptr(3),
+        fq->get_input_node_shared_ptr(4),
+        as_type_ptr<opset1::FakeQuantize>(fq)->get_levels());
+    fq2->set_friendly_name("concat_fq2");
+
+    auto newConcat = concat->clone_with_new_inputs({ fq1->output(0), fq2->output(0) });
+    newConcat->set_friendly_name(concat->get_friendly_name());
```
[the remark this answers is not in the document]
Comment (eshoguli) [reply inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp:98]: `NetworkHelper::copyInfo` is skipped

File: inference-engine/tests/functional/inference_engine/lp_transformations/move_fake_quantize_transformation.cpp [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+#include <low_precision/low_precision.hpp>
+
+#include "low_precision/move_fake_quantize.hpp"
+#include <low_precision/fake_quantize_decomposition.hpp>
+#include <low_precision/rt_info/precision_preserved_attribute.hpp>
+#include <low_precision/align_quantization_parameters.hpp>
+#include <low_precision/fuse_subtract_to_fake_quantize.hpp>
+#include <low_precision/fuse_multiply_to_fake_quantize.hpp>
+#include <low_precision/markup_can_be_quantized.hpp>
+#include <low_precision/markup_per_tensor_quantization.hpp>
```
Comment (eshoguli) [inline inference-engine/tests/functional/inference_engine/lp_transformations/move_fake_quantize_transformation.cpp:29]: please, remove not used headers in the file

**Comment by eshoguli** [comment]
tests review is not completed yet

File: inference-engine/tests/functional/inference_engine/lp_transformations/move_fake_quantize_transformation.cpp [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+        auto quantizationRestrictions = testValues.multiChannels ?
+            std::vector<ngraph::pass::low_precision::OperationPerTensorQuantizationRestriction>() :
+            std::vector<ngraph::pass::low_precision::OperationPerTensorQuantizationRestriction>({
+                ngraph::pass::low_precision::OperationPerTensorQuantizationRestriction::create<ngraph::opset1::AvgPool>()
+                });
+
+        const auto params = TestTransformationParams::toParams(testValues.params);
+        SimpleLowPrecisionTransformer transformer(supportedPrecisionsOnActivation, quantizationRestrictions);
+        transformer.commonGraphRewrite->add_matcher<ngraph::pass::low_precision::MoveFakeQuantize>(params);
+        transformer.transform(actualFunction);
```
Comment (eshoguli) [inline inference-engine/tests/functional/inference_engine/lp_transformations/move_fake_quantize_transformation.cpp:183]: Use implemented transformation only:
```suggestion
        ov::pass::Manager manager;
        manager.register_pass<ngraph::pass::low_precision::MoveFakeQuantize>(params);
        manager.run_passes(actualFunction);
```

File: inference-engine/tests/functional/inference_engine/lp_transformations/move_fake_quantize_transformation.cpp [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+            "axis_" << testValues.axis << "_" <<
+            testValues.actual << "_" <<
+            testValues.result << "_";
+        return result.str();
+    }
+};
+
+TEST_P(MoveFakeQuantizeTransformation, CompareFunctions) {
+    actualFunction->validate_nodes_and_infer_types();
+    auto res = compare_functions(referenceFunction, actualFunction, true, true, false, true, false);
```
Comment (eshoguli) [inline inference-engine/tests/functional/inference_engine/lp_transformations/move_fake_quantize_transformation.cpp:232]:
```suggestion
    auto res = compare_functions(referenceFunction, actualFunction, true, true, true, true, true);
```

File: inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        return transform(*context, m);
+    };
+
+    auto m = std::make_shared<ngraph::pattern::Matcher>(
+        std::make_shared<pattern::op::Or>(OutputVector{fq, fq_with_operation}),
+        "MoveFakeQuantize");
+    this->register_matcher(m, callback);
+}
+
+bool MoveFakeQuantize::transform(TransformationContext& context, ngraph::pattern::Matcher& m) {
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp:56]: if LPT `ConcatTransformation` doesn't handle this `Concat`, then `MoveFakeQuantizeTransformation` should not handle too. You need to make existing `Concat::canBeTransformed` static (for example: `Concat::canBeTransformedStatic`) and reuse it in the `MoveFakeQuantizeTransformation`. Don't forget negative tests cases in functional & plugin tests, please.

**Review by eshoguli — CHANGES_REQUESTED** [review]
please, fix comments

File: inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    std::vector<std::shared_ptr<ngraph::Node>> fqs;
+    size_t input_size = concat->get_input_size();
+    for (size_t i{ 0 }; i < input_size; ++i) {
+        std::shared_ptr<ngraph::Node> fq_input;
+        if (only_concat) {
+            fq_input = concat->get_input_node_shared_ptr(i);
+        } else {
+            auto input = concat->get_input_node_shared_ptr(i);
+            fq_input = std::make_shared<ngraph::opset1::Relu>(input);
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp:81]: minor: `clone_wiith_new_inputs` is more preferable

File: inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+            fq_input->set_friendly_name(relu_original_name + std::to_string(i + 1));
```
Comment (eshoguli) [inline inference-engine/src/low_precision_transformations/src/move_fake_quantize.cpp:82]: minor: `fq_input->set_friendly_name(relu_original_name + "_" + std::to_string(i + 1));`

**Review by eshoguli — APPROVED** [review]
please fix minor comments in D/DQ support PR

**Pull request merged**